# **Part 1: Tasks.**
The following paragraphs provide the names of each task, along with a brief description and the commands used in the Bash terminal on Linux Virtual Machine in Github Codespaces to complete them.

## 1.1 Create Directory Structure.
As per instructions created a directory named **data** at the root of the repository **Computer-Infrastructure** ```➜ /workspaces/Computer-Infrastructure/```and inside **data**, two subdirectories: **timestamps** and **weather**. 
Commands:

```bash
mkdir -p data/timestamps
mkdir -p data/weather

```-p``` ensures that the parent directory (data) is created if it doesn't already exist.

## 1.2 Timestamps.
Used the ```date``` command to output the current date and time inside the **data/timestamps** folder, appending the output to a file named **now.txt**. In order to append the file the ```>>``` operator was used. Repeated this step ten times, then verfied contents of **now.txt** using the ```more``` command:

```bash
date >> now.txt
more now.txt

## 1.3 Formatting Timestamps.
```date``` command was run again, but this time the output format was YYYYmmdd_HHMMSS (e.g.,20241127_220039, 22:00:39 on Nov.27th 2024) and was appended to a file named **formatted.txt**. Commands:
```bash
date +"%Y%m%d_%H%M%S" >> formatted.txt
man date

```man date``` displays and briefly describes all commands that can be used with ```date``` . For example:
- %F = full date; same as %Y-%m-%d  
- %T = time; same as %H:%M:%S  
  
therefore
```bash
date +"%F_%T" >> formatted.txt  
```  
will append this time format to **formatted.txt**: 2024-12-02_19:42:50 (time of writing).

## 1.4 Create Timestamped Files.
In order to create timestamped files without using ```>>``` redirection embedded ```date```command in backticks:
```bash
touch `date +"%Y%m%d_%H%M%S"`.txt   

## 1.5 Download Today's Weather Data.
In this task it was requested to download the latest weather data for the Athenry weather station URL: https://prodapi.metweb.ie/observations/athenry/today from Met Eireann using wget. The -O <filename> option had to be used to save the file as weather.json:
```bash
wget -O weather.json https://prodapi.metweb.ie/observations/athenry/today
```
```wget``` is a command-line utility for downloading files from the web.  
```-O``` stands for output and specifies the name of the file where the downloaded content should be saved.


## 1.6 Timestamp the Data.
The command from previous task had to be modified to save the downloaded file with a timestamped name in the format YYYYmmdd_HHMMSS.json. This was achieved by replacing ```weather.json``` from above command with ```"$(date +"%Y%m%d_%H%M%S").json"```:

```bash
wget -O "$(date +"%Y%m%d_%H%M%S").json" https://prodapi.metweb.ie/observations/athenry/today


## 1.7 Write the script.
A bash script called ```weather.sh``` was created at the root of the repository **Computer-Infrastructure** ```➜ /workspaces/Computer-Infrastructure/```. This script had to automate the process from 1.6, saving the weather data to the data/weather directory. This task was completed by running a series of commands:  
1. ```touch weather.sh``` to create ```weather.sh```.   
2. ```cat > weather.sh``` interactive method for adding content. 
3. Subsequently entered the following lines, then Ctrl + D to save and finish the operation:
```bash
#! /bin/bash

 date
 echo "Downloading weather data"
 wget -O data/weather/`date +"%Y%m%d_%H%M%S.json"` https://prodapi.metweb.ie/observations/athenry/today
 echo "Weather data downloaded"
 date
 ```
4. ```chmod +x weather.sh``` to make the file executable.






# **Part 2: Weather Analysis**
### A brief glance at Athenry's weather data as per metweb.ie.
In this section, the knowledge gained from the tasks in Part 1 is applied to a real-world scenario.
## 2.1 Import data frame.

In [26]:
# Data frames
import pandas as pd

## 2.2 Collecting the data.

```bash
#! /bin/bash

 date
 echo "Downloading weather data"
 wget -O data/weather/`date +"%Y%m%d_%H%M%S.json"` https://prodapi.metweb.ie/observations/athenry/today
 echo "Weather data downloaded"
 date
```

Data is collected via metweb.ie's API and stored as JSON file in the following format YearMonthDay_HourMinuteSecond (e.g 20241127_220039, collected on Nov.27th 2024 at 22:00 39)

## 2.3 Anayzing the data.
### 2.3.1 Using Pandas to arrange and output the data for lean usability:

In [27]:
# Read the data.
df = pd.read_json('data/weather/20241127_220039.json')

### 2.3.2 Let's have a look at the head of the data (5 top lines):

In [28]:
# Show.
df.head()

name  temperature symbol weatherDescription              text  \
0  Athenry            0    15n         Fog / Mist            "Mist"   
1  Athenry            0    15n         Fog / Mist  "Fog thickening"   
2  Athenry           -1    15n         Fog / Mist             "Fog"   
3  Athenry           -2    15n         Fog / Mist  "Fog thickening"   
4  Athenry           -3    15n         Fog / Mist             "Fog"   

   windSpeed windGust cardinalWindDirection  windDirection  humidity  \
0          6        -                    NE             45        98   
1          6        -                    NE             45        99   
2          2        -                    NE             45        98   
3          6        -                    NW            315        98   
4          6        -                     N              0        97   

   rainfall  pressure    dayName       date reportTime  
0         0      1015  Wednesday 2024-11-27      00:00  
1         0      1015  Wednesday 2024-11-27      01:00  
2         0      1015  Wednesday 2024-11-27      02:00  
3         0      1015  Wednesday 2024-11-27      03:00  
4         0      1015  Wednesday 2024-11-27      04:00

### 2.3.3 Summary of descriptive statistics:

In [29]:
# Show.
df.describe()

temperature  windSpeed  windDirection  humidity  rainfall     pressure  \
count    22.000000  22.000000      22.000000   22.0000      22.0    22.000000   
mean     -1.363636   4.636364      98.181818   98.0000       0.0  1018.136364   
min      -3.000000   2.000000       0.000000   96.0000       0.0  1015.000000   
25%      -2.000000   2.000000       0.000000   97.2500       0.0  1015.250000   
50%      -1.000000   4.000000      67.500000   98.0000       0.0  1018.500000   
75%      -0.250000   6.000000     123.750000   99.0000       0.0  1020.000000   
max       0.000000   9.000000     315.000000   99.0000       0.0  1022.000000   
std       1.135801   2.300009     113.363569    0.9759       0.0     2.550359   

                      date  
count                   22  
mean   2024-11-27 00:00:00  
min    2024-11-27 00:00:00  
25%    2024-11-27 00:00:00  
50%    2024-11-27 00:00:00  
75%    2024-11-27 00:00:00  
max    2024-11-27 00:00:00  
std                    NaN

- count: Number of non-missing values.
- mean: Average value.
- std: Standard deviation.
- min: Minimum value.
- 25%: First quartile (25th percentile).
- 50%: Median (50th percentile).
- 75%: Third quartile (75th percentile).
- max: Maximum value.

### 2.3.4 Overview of the DataFrame's structure and contents:

In [30]:
print(df.info())


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 22 entries, 0 to 21
Data columns (total 15 columns):
 #   Column                 Non-Null Count  Dtype         
---  ------                 --------------  -----         
 0   name                   22 non-null     object        
 1   temperature            22 non-null     int64         
 2   symbol                 22 non-null     object        
 3   weatherDescription     22 non-null     object        
 4   text                   22 non-null     object        
 5   windSpeed              22 non-null     int64         
 6   windGust               22 non-null     object        
 7   cardinalWindDirection  22 non-null     object        
 8   windDirection          22 non-null     int64         
 9   humidity               22 non-null     int64         
 10  rainfall               22 non-null     int64         
 11  pressure               22 non-null     int64         
 12  dayName                22 non-null     object        
 13  date   

We learn that the data set has 14 columns with 22 entries each which are data type object(8), int64(6) as well as datetime64(1)
Interesting is also the display of total memory usage of the DataFrame: 2.7+ KB.

Let's print the data:

In [ ]:
pd.set_option('display.width', 1000) # Printing column next to each other
print(df)

       name  temperature symbol weatherDescription              text  windSpeed windGust cardinalWindDirection  windDirection  humidity  rainfall  pressure    dayName       date reportTime
0   Athenry            0    15n         Fog / Mist            "Mist"          6        -                    NE             45        98         0      1015  Wednesday 2024-11-27      00:00
1   Athenry            0    15n         Fog / Mist  "Fog thickening"          6        -                    NE             45        99         0      1015  Wednesday 2024-11-27      01:00
2   Athenry           -1    15n         Fog / Mist             "Fog"          2        -                    NE             45        98         0      1015  Wednesday 2024-11-27      02:00
3   Athenry           -2    15n         Fog / Mist  "Fog thickening"          6        -                    NW            315        98         0      1015  Wednesday 2024-11-27      03:00
4   Athenry           -3    15n         Fog / Mist     

### 2.4 Summary
This file contains a collection of hourly weather observations from the synoptic station in Athenry, Co. Galway, for November 27th. The data, collected at 22:00 local time, provides 22 entries for each of the following variables: station name, dry bulb temperature (in whole degrees Celsius), weather description, windspeed (in knots), cardinal wind direction, relative humidity (percentage), rainfall (in millimeters), mean sea level pressure (in millibars), day of the week, date, and time of observation. November 27th was notably cold, with temperatures ranging from 0°C to below freezing. The day remained dry with no recorded rainfall but experienced fog to thick fog. Weather conditions were otherwise stable due to consistently high air pressure exceeding 1015 millibars throughout the day.

### End.